pkg/southbound/gnmi/conn_manager.go

```diff
@@ -0,0 +1,191 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gnmi
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc"
+
+	"google.golang.org/grpc/connectivity"
+
+	topoapi "github.com/onosproject/onos-api/go/onos/topo"
+
+	"github.com/onosproject/onos-lib-go/pkg/errors"
+)
+
+// ConnManager gNMI connection manager
+type ConnManager interface {
+	Get(ctx context.Context, id ConnID) (Conn, error)
+	List(ctx context.Context) ([]Conn, error)
+	Watch(ctx context.Context, ch chan<- Conn) error
+	Connect(ctx context.Context, target *topoapi.Object) (Conn, error)
+	remove(connID ConnID) error
+}
+
+// NewConnManager creates a new gNMI connection manager
+func NewConnManager() ConnManager {
+	mgr := &connManager{
+		conns:       make(map[ConnID]Conn),
+		clientConns: make(map[topoapi.ID]*grpc.ClientConn),
+		eventCh:     make(chan Conn),
+	}
+	go mgr.processEvents()
+	return mgr
+}
+
+type connManager struct {
+	conns       map[ConnID]Conn
+	clientConns map[topoapi.ID]*grpc.ClientConn
+	connsMu     sync.RWMutex
+	watchers    []chan<- Conn
+	watchersMu  sync.RWMutex
+	eventCh     chan Conn
+}
+
+// Connect connecting to a gNMI target and adding a new gNMI connection
+func (m *connManager) Connect(ctx context.Context, target *topoapi.Object) (Conn, error) {
+	log.Infof("Connecting to the gNMI target: %s", target.ID)
+	conn, err := connect(target)
+	if err != nil {
+		log.Errorf("Failed to connect to the gNMI target: %s", target.ID)
+		return nil, err
+	}
+
+	log.Infof("Adding gNMI connection %s for the target %s", conn.ID(), target.ID)
+	m.connsMu.Lock()
+	_, ok := m.conns[conn.ID()]
+	if ok {
+		m.connsMu.Unlock()
+		return nil, errors.NewAlreadyExists("gNMI connection %s already exists", conn.ID())
+	}
+
+	m.conns[conn.ID()] = conn
+	m.clientConns[target.ID] = conn.clientConn()
+	m.connsMu.Unlock()
+	m.eventCh <- conn
+
+	go func() {
+		clientConn := conn.clientConn()
+		state := clientConn.GetState()
+		for state != connectivity.Shutdown && clientConn.WaitForStateChange(ctx, state) {
+			state := clientConn.GetState()
+			log.Infof("Current state of the gNMI connection %s: is %s", conn.ID(), state.String())
+			if state == connectivity.TransientFailure {
+				log.Infof("Closing the gNMI connection for target %s:%s", conn.ID(), target.ID)
+				m.connsMu.Lock()
+				if conn, ok := m.conns[conn.ID()]; ok {
+					delete(m.conns, conn.ID())
+					m.eventCh <- conn
+				}
+				m.connsMu.Unlock()
+				break
+			}
+			if state == connectivity.Ready {
+				m.connsMu.Lock()
+				if _, ok := m.conns[conn.ID()]; !ok {
+					conn := newConn(target)
```

Comment:
Remember even though we’re creating a separate `Conn` instance for each physical connection to the target, they should all share the same `grpc.ClientConn`. The `newConn` function doesn’t allow us to pass in a `grpc.ClientConn` to use, so I assume it’s creating a new one internally? I think the `grpc.ClientConn` needs to be created once here in the `Connect' method and passed into `newConn` each time we create a new `Conn` instance for the connection.